#162 API for retrieving info from the Adelphi cluster

grighetto opened the pull request:
TO BE REVIEWED AFTER #166

grighetto commented on api/src/api/index.js:
```diff
@@ -0,0 +1,66 @@
+  };
+  
+  const errorCount = checkErrors(result);
+  const code = errorCount > 0 ? 500 : 200;
```
If there were any integrity failures (gemini or cassandra_diff), the HTTP status code will be 500, otherwise it will be 200.
This helps with clients like `curl` that will return a non-zero exit code if the HTTP status is not in the 2xx range.

grighetto commented on api/src/api/index.js:
```diff
@@ -0,0 +1,66 @@
+  console.log("Data path: " + DATA_PATH);
+
+  const result = {
+    gemini: geminiSummary(readDataFile(DATA_PATH + "/gemini/results.json")),
```
> Provide API endpoints that will compress and respond with the full set of results files?

Totally agree, that's actually coming up in a new PR soon, I just didn't want to include everything in a single shot so we can agree on the foundation first.

I just didn't think about removing the old, file-based, results server, but we can probably do it once we have the summary API + ability to download everything. +1
That will simplify the nginx config because we won't have to do any proxy forwarding. The API will be the main entrypoint.

grighetto commented on api/src/server.js:
```diff
@@ -0,0 +1,8 @@
+const PORT = process.env.PORT || 3000;
+
+api.listen(PORT, () => {
+  console.log(`Adelphi API listening at http://localhost:${PORT}`)
```
The same way we access the results server right now with `kubectl port-foward`.
So if you hit the /api/ path, it will proxy to the node server internally:

https://github.com/datastax/adelphi/pull/167/files#diff-9d4b4d1d3a7b9885d6a08c0673d7bfdf12d031ceb458c08038c0cbdc32f21951R17

grighetto commented on api/src/api/index.js:
```diff
@@ -0,0 +1,66 @@
+  generated results files.
+*/
+export function resultsSummary(req, res) {
+  console.log("GET /api/results/summary");
```
Yes, we should definitely do that. +1
I captured that in https://github.com/datastax/adelphi/issues/172 so we can split that work and unblock this PR.

jdonenine reviewed (approved):
I got a chance to test out the workflow with the newly added APIs, that's a good add for ease of use of accessing results.  Looking forward to seeing what else we can do on that front with this here as a base.

I can go ahead and approve this, but the new ITs are failing on this PR, so you'll want to look at that @grighetto .  Since they are so new I'm guessing there are some kinks to be worked out there.

grighetto commented:
> but the new ITs are failing on this PR, so you'll want to look at that

@jdonenine I did some investigation and there are mainly two flaky cases:
1- The create-schema step fails because it cannot authenticate to one of the clusters. Even though we check for the Ready state of the C* pods, it seems that might not be enough in constrained environments such as GHA. To fix this in Adelphi we can test the cqlsh connection before we try to create the schema.
2- It cannot start the results-server in the last step due to some volume affinity conditions. I'll punt on this for now because it seems it only rarely happens on GHA.

I'll capture these issues into separate tickets so we can track them and split them from this PR.

Note: The IT tests are failing on C* *trunk* due to https://github.com/k8ssandra/management-api-for-apache-cassandra/issues/122, but I added a new GHA workflow to confirm it works against C* 4.0.
